File: go/backtracking/78_subsets.go
package backtracking

import "slices"

// 子集
func subsets(nums []int) [][]int {

	res := make([][]int, 0)
	path := make([]int, 0)

	var backtrack func(startIdx int)

	backtrack = func(startIdx int) {
		p := slices.Clone(path)
		res = append(res, p)

		for i := startIdx; i < len(nums); i++ {
			path = append(path, nums[i])

			backtrack(i + 1)
			path = path[:len(path)-1]
		}

	}

	backtrack(0)
	return res
}

File: rust/src/backtracking/_78_subsets.rs
/// 子集
pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let mut res = vec![];
    let mut path = vec![];

    fn backtrack(start_idx: usize, nums: &[i32], res: &mut Vec<Vec<i32>>, path: &mut Vec<i32>) {
        res.push(path.clone());

        for i in start_idx..nums.len() {
            path.push(nums[i]);
            backtrack(i + 1, nums, res, path);
            path.pop();
        }
    }

    backtrack(0, &nums, &mut res, &mut path);
    res
}
